File: app/src/main/java/com/cargopull/executor_driver/gateway/Mapper.java
package com.cargopull.executor_driver.gateway;

import androidx.annotation.NonNull;

/**
 * Однонаправленный преобразователь данных из формы {@link F} в форму {@link T}.
 *
 * @param <F> тип исходных данных
 * @param <T> тип выходных данных
 */
public interface Mapper<F, T> {

  /**
   * Преобразует отдельный элемент типа {@link F} в тип {@link T}.
   *
   * @param from исходный элемент
   * @return преобразованный элемент
   * @throws Exception если возникает какое-либо несовместимое с преобразованием обстоятельство
   */
  @NonNull
  T map(@NonNull F from) throws Exception;
}

File: app/src/main/java/com/cargopull/executor_driver/gateway/Mapper.kt
package com.cargopull.executor_driver.gateway

/**
 * Однонаправленный преобразователь данных из формы [F] в форму [T].
 *
 * @param <F> тип исходных данных
 * @param <T> тип выходных данных
</T></F> */
interface Mapper<F, T> {

    /**
     * Преобразует отдельный элемент типа [F] в тип [T].
     *
     * @param from исходный элемент
     * @return преобразованный элемент
     * @throws Exception если возникает какое-либо несовместимое с преобразованием обстоятельство
     */
    @Throws(Exception::class)
    fun map(from: F): T
}

class MirrorMapper<T> : Mapper<T, T> {
    override fun map(from: T): T = from

}

class ConstantMapper<T, D>(private val value: D) : Mapper<T, D> {
    override fun map(from: T): D = value

}